Unified metadata for compilation files (or no more capitalize_ascii)

Octachron opened the pull request:
This PR is an alternative implementation of a slice of #11736 that is entirely focused on the handling of metadata (module name and source file currently) for compilation files (either compilation targets or source files) in order to avoid the proliferation of `String.{uncapitalize_ascii,capitalize_ascii}` within the compiler codebase.

In brief, this PR makes sure that we define the transformations from compilation filenames to module names in one place, and the same for the reverse transform. 
(However, dependencies made it easier to define the two functions in two different places).

After this PRs, the only calls to `String.capitalize_ascii` in the compiler happen in error message printers (typically to capitalize either `first` or `second`) and are only an obstacle to the internationalization of the compiler messages.

The PR goes one step further and ensures that the module name associated to a compilation file is computed once by tracking file metadata across change of file extensions.

With this change, it is quite straightforward to check that the compiler only tries to derive filenames from module names when doing lookup for `cmo`, `cmx` or `cmi` file through `Load_path`.

This suggest to me that a good solution for the compilation artifact ambiguity (which files between `Foo.cmi` and `foo.cmi` should provide the module `Foo`?) is to emit an error if the same directory contains distinct filename with the same normalization as suggested by @dbuenzli in #11736.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+(**  A value [file] of type {!t} records
+      - the module name associated to the file
+      - the target filename or prefix
+      - the source file for compilation file if it exists
```
Two remarks on this type documentation:
- the explanations do not let me understand what `'source` and `´target` are
- I am used to documentation starting with "A value of type [t] ...", but you chose to suggest a typical variable name `file` for values of our type, which I suppose has a special significance. But I am confused as to why a `Unit_info.t` would be named `file` (and not, say, `unit` or `unit_info` or `file_info`). I guess that there is some subtlety here that could be explained more explicitly.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+  Maybe more importantly, this module provides functions for deriving module
+  names from strings or filenames.
+*)
```
Would it make sense to give some idiomatic examples of functions in this module, as a way to illustrate this high-level description and make some things in the documentation clearer?

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+type source_file = (string, unit) t
+type target_prefix = (string, unit) t
+(** The prefix of a target file with its module name and source file *)
```
At this point I don't understand why `target_prefix` has `source:string` and `target:unit`, I would have expected a use of the `target` type for a target prefix? (Or maybe `target_prefix` is a confusing name and some other is meant, for example "working directory"? At this point I realize that I am not sure what you mean by "target prefix".)

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+type target = string any_target
+type target_only = unit any_target
+(** A target file without a source file: this typically happen when packing *)
```
I am not sure what those synonyms are good for, except for `target_only` which I understand. Is `target` used anywhere?

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+(** [modname_from_source filename] is [modulize stem] where [stem] is the
+    basename of the filename [file] stripped from all its extensions.*)
+val modname_from_source: string -> string
```
This sort of functions would be easier to read with "silent" type abbreviations just for this module:

```
type path = string
type filepath = string
type dirpath = string
type modname = string
```

gasche replied:
(I wouldn't mind `type modname = Modname of string [@@unboxed]` either, but we could start with a synonym to avoid caller-side changes.)

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+(** [is_unit_name ~strict name] returns true only if [name] can be used as a
+    correct module name. The [strict] mode enforces that the first character is
+    an ascii capital letter. *)
+val is_unit_name : strict:bool -> string -> bool
```
It is not clear to me why we have this function. Is it possible for `modulize` to return some output that fails this check? (Should this be documented?). I would make the guess that a good API would be a partial function `modulize : string -> (string, warning) result` or something, but that having a total function made it easier to integrate in the compiler codebase.

Octachron replied:
This check implements the warning raised when compiling `ocamlopt 1.ml -o <any>`, which creates a compilation unit with the module name `1`.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+    [modname_from_source filename]. *)
+val from_target: string -> target_only
+
+(** [from_target_prefix ~source_file target_prefix] aasociates both the
```
associates

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+(** [from_source filename] associates the module name [modname_from_source
+    filename] to the source file [filename]. *)
+val from_source: string -> source_file
```
These docstrings are a bit confusing because the type is `path -> source_file`, but the description is about module names. I wonder if a more abstract description would make sense, for example:

```ocaml
(** [from_source filename] is the unit information for the source file [filename]. *)
```

gasche commented on parsing/unit_info.ml:
```diff
@@ -0,0 +1,151 @@
+  let modname = modname_from_source filename in
+  Target_only { modname; filename }
+
+let target_from_prefix ext (type a b) (p: (a,b) t): (a,string) t =
```
The name is a bit strange as only the first two cases seem to use a prefix, and also the name `target_from_prefix` is uncomfortably close to `from_target_prefix`. Maybe `target_with_extension`?

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+(** [cmi_uncap u] finds in the load_path a name matching the module name
+    [modname u]. *)
+val cmi_uncap: _ t -> target_only
```
I would name this `find_cmi_uncap` to keep in mind that there is a non-trivial lookup logic behind it (with a performance cost, etc.).

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+
+(** [from_target filename] associates the target [filename] to the module name
+    [modname_from_source filename]. *)
+val from_target: string -> target_only
```
I was a bit confused as to what `from_target` means here. (My first guess was that the name was derived from the `-o <target>` argument passed by the user.) In fact, looking at the callsites of `from_target`, it looks like they are called on paths that are not the *target* of the current build command, but a *source* of the build command that happens to be an object file (typically a `.cmo` file during linking).

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,136 @@
+val cmx_name: _ t -> string
+val cmo_name: _ t -> string
+val obj_name: _ t -> string
+val annot_name: _ t -> string
```
For me `name` is not so clear, I prefer `path` (filesystem path) when applicable.

gasche replied:
On reading the API, it is unclear why some extensions such has `.annot` have both `annot` and `annot_name`, `.cmi` has no `cmi_name` and `.cmx` has no `cmx` function. I would find it more regular to have only functions returning a `('a, string) t` value for those extensions, and then a `target_path : (_, string) t -> path` function to get the output path.

gasche reviewed (commented):
In theory this is a very nice change, and the callsites are all improved. I am broadly in favor of merging this work.

In practice I read `unit_info.mli` without looking at any of the other parts of the PR, and I found it fairly confusing. I think that if we want other people to work on that part of the compiler codebase, we should try to have (besides the nice refactoring) an interface that people can understand. I did a first round of feedback focusing on my various sources of confusion on unit_info.mli.

gasche commented:
Some more vague comments:

1. I think that we have different things in mind about the name "target". For me a target is an *output* of the current compiler build action. I have the impression that what you call "target" are files that are built by *some* compiler invocation, the current one or a previous one. Those I call "build artifacts" or "object files".

2. I am unsure what is called a "prefix" here. My current guess is that it denotes both a working directory and what you call a "stem", an extensionless basename fragment that is used to derive the module name. (In contrast, for opam or other package management tools the "prefix" is only a directory path.) I wonder if systems people have a more standard name for "stem", but I am fine with this choice.

3. `Unit_info.t` pairs information about compiler objects (what path to write them to) and "provenance" information on the compiler inputs this information was derived from. Some operations of the API change the compiler object (eg. "move from a .cmo to the corresponding .cmi") and keep the same provenance. I wondered if the API or at least the implementation could be clarified by being explicit about this product structure: have a part of the data that only denotes a "compilation unit" (a module name, an optional source filename, a filesystem prefix), and a separate part that denotes one compiler object among the many that can be derived from it.

    (Without looking at the code in details, my intuition would be that the object-specific part does not really need to be stored along the unit information: functions that take a `cmx_name` as input, and currently get passed an `Unit_info.cmx_name unit` could be rewritten to take `unit` directly and ask for its cmx path at each use-site. The object-specific information, probably just an extension, needs to be passed only on functions that can work with several object file extensions in a generic way, and those are probably very rare and could take the extension as an extra argument separate from the unit info or something. But that is a more invasive change on the caller site, so why not have this intermediate form where we pair both together to stick closer to the previous path-only approach.)

Note: remarks (1) and (2) above are about misunderstandings on what we mean by specific names; providing examples in the beginning of the .mli documentation would reduce the opportunities for such misunderstandings. (We could still disagree on whether the names are the right ones, but at least the reader would quickly be sure of what they mean.)

gasche commented:
@Octachron and myself discussed this again yesterday evening. For reference here is my new understanding of the concepts in unit_info.mli:

- The type `source_file` represents (the filesystem paths to) a source compilation unit, that is basically a pair of a .ml file and a .mli file (or only some of them sometimes, I guess); I would call this `source_unit` or `source_unit_info`.
- The type `target_prefix` represents (the filesystem paths to) an "compilation unit object" (or *compiled* compilation unit), that is a bunch of compiled object files (.cmi, possibly .cmt, .cmo or .cmx, etc.), *along* with information about its source compilation unit. I think we could call this a `compilation_unit`, or `compilation_unit_info`, or just `t` as it is the central type of the module.
- The type `'a any_target` represents (the filesystem path to) a single object file that is part of an object compilation unit, derived from a compilation unit object. If `'a` is `string` (type `target`), then it knows of an existing a source compilation unit, if `'a` is `unit` (type `target_only`) then it was derived from an object file and has no source compilation unit attached.

I asked @Octachron whether we could differentiate "whole units" and "single files" with separate types:
- a type `source_unit` for source compilation units
- a type `'a t` for compiled compilation units that also carry a `'a`: `source_unit t` has source information, `unit t` has only object information
- a type `'a object` or `'a artifact` for a single object file derived from a `'a t`

That interface would be less polymorphic than the current one (where some helper functions can take any `(_, _) t` as input, and would have to be duplicated to work on the different types), and it is not clear whether it would be a problem in practice or not. I think that @Octachron wants to hack on it a bit to see what would work and what would not.

Octachron commented:
I have removed the unified GADTs and split it into an `Unit_info.t` and `Unit_info.artifact` types.

The other cases were not strictly necessary : this simplified version only loses the information that only artifact metadata derived from `Unit_info.t` contains source file information.

This required a few function duplication, but the resulting interface should be far easier to read.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,140 @@
+    - the module name associated to the unit
+    - the filename prefix (dirname + basename with all extensions stripped)
+      for compilation artifacts
+    - the source file
```
A concrete example would help the reader.

```
- the module name associated to the unit; for example "Mylib_Foo"
- the filename prefix (...); for example "_build/mylib/Mylib__Foo"
- the source file; for example "mylib/foo.ml"
```

(I think of compilation units as typically having two source files, the .ml and the .mli, so I am not sure what "the source file" means.)

Octachron replied:
I have added an example. The source file refers to the lone input source file, since the compilation pipeline proceeds file-by-file.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,140 @@
+(**  Metadata for a single compilation artifact:
+    - the module name associated to the artifact
+    - the filesystem path
+    - the source file for compilation file if it exists
```
```
Metadata for a single compilation artifact, for example a .cmi or .cmx file:
- the module name associated to the artifact; for example "Foo"
- the filesystem path; for example "_build/src/foo.cmx"
- the source file if it is known; for example "src/foo.ml"
```

(again I am not sure what "the source file" means)

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,140 @@
+val source_file: t -> string
+
+(** [artifact_source_file a] is the source file of [a] if it exists. *)
+val artifact_source_file: artifact -> string option
```
... at the risk of being nitpicky, I wonder how an `Artifact` submodule would feel.

Octachron replied:
I have added such submodule, since it helps a lot distinguish the field of the `Artifact.t` and `Unit_info.t` types.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,140 @@
+
+(** [normalized_cmi u] finds in the load_path a file matching the module name
+    [modname u]. *)
+val normalized_cmi: t -> artifact
```
If it raises Not_found, I think it should be named `find_normalized_cmi`,  and the exception should be documented.

gasche commented on parsing/unit_info.mli:
```diff
@@ -0,0 +1,140 @@
+
+(** [artifact filename] reconstruct the module name
+    [modname_from_source filename] associated to the artifact [filename]. *)
+val artifact: string -> artifact
```
This reads weird because all other `Unit_info.<noun>` function are used to project information out of the datatypes of this module. Maybe `Unit_info.Artifact.make` or `Unit_info.Artifact.from_path`?

gasche commented on debugger/command_line.ml:
```diff
@@ -153,11 +153,7 @@ let module_of_longident id =
-      String.capitalize_ascii(if Filename.check_suffix m ".ml"
-                              then Filename.chop_suffix m ".ml"
-                              else m)
+  | Some m -> Unit_info.modname_from_source m
```
The code comment was previously clear that all of `foo.ml`, `foo` and `Foo` could be given -- intentionally, I suspect. I don't know if the new code allows all these options -- it is unclear from the code and, if it is the case, should be documented explicitly.

In practice it looks like all callers of the `convert_module` function are of the form `convert_module (module_of_longident foo)`. So I think that instead of your code, that makes thing morally wrong (but maybe technically preserves the behavior), we should simply have a `convert_longident` function that does the right thing without invoking functions meant for file paths.

Octachron replied:
There was a bug here, because `m` might be a module path. This has been fixed to preserve the existing behavior.

gasche commented on driver/compenv.mli:
```diff
@@ -46,13 +44,6 @@ type readenv_position =
-val is_unit_name : string -> bool
-(* [check_unit_name ppf filename name] prints a warning in [filename]
-   on [ppf] if [name] should not be used as a module name. *)
-val check_unit_name : string -> string -> unit
```
I wonder if we should keep these functions around for compiler-libs-users compatibility: it is basically free to do so by defining them in terms of Unit_info operations. (Looking for `Compenv.is_unit_name` for example on github finds exactly one hit in the Malfunction codebase.)

stedolan replied:
Came across this comment when updating said line!

I don't think it's worth keeping these things around in compiler-libs. Malfunction (and anything else heavily using compiler-libs) already has to deal with these things changing from release to release, and adding one more line of conditional compilation isn't a huge burden.

gasche commented on driver/compile_common.mli:
```diff
@@ -17,9 +17,7 @@
-  source_file : string;
-  module_name : string;
-  output_prefix : string;
+  target : Unit_info.t;
```
(It is good that `with_info` below was kept as-is for backward compatibility for compiler-libs users.)

gasche commented on driver/compile_common.mli:
```diff
@@ -77,13 +75,3 @@ val typecheck_impl : info -> Parsetree.structure -> Typedtree.implementation
-val cmo : info -> string
-val cmx : info -> string
-val obj : info -> string
-val annot : info -> string
```
A brief Github code search seems to suggest that no one was using those.

gasche commented on driver/optcompile.ml:
```diff
@@ -41,7 +42,7 @@ let flambda i backend Typedtree.{structure; coercion; _} =
   (structure, coercion)
   |> Profile.(record transl)
-      (Translmod.transl_implementation_flambda i.module_name)
+      (Translmod.transl_implementation_flambda (Unit_info.artifact_modname cmx))
```
Is there a reason for using this instead of `Unit_info.modname i.target`?
If you used the latter you wouldn't need to define `cmx` early on, which is not particularly natural in this function, you could do just `Unit_info.(filename (Artifact.cmx i.target))`.

gasche commented on driver/optcompile.ml:
```diff
@@ -64,19 +65,20 @@ let flambda i backend Typedtree.{structure; coercion; _} =
 
 
 let clambda i backend Typedtree.{structure; coercion; _} =
+  let cmx = Unit_info.cmx i.target in
```
(meh again)

gasche commented on driver/optcompile.ml:
```diff
@@ -64,19 +65,20 @@ let flambda i backend Typedtree.{structure; coercion; _} =
           Asmgen.compile_implementation
             ~backend
-            ~prefixname:i.output_prefix
+            ~prefixname:(Unit_info.prefix i.target)
```
Each call to `compile_implementation` needs to be modified. Maybe it could even be changed to take a `unit_info` instead of a `prefixname`, to make caller-side changes simpler?

Octachron replied:
I think it is better to drop the information which is no longer used in this case.

gasche commented on driver/optcompile.ml:
```diff
@@ -87,21 +89,23 @@ let clambda i backend Typedtree.{structure; coercion; _} =
-  Asmgen.compile_implementation_linear i.output_prefix ~progname:i.source_file
+  Compilenv.reset ?packname:!Clflags.for_package (Unit_info.modname i.target);
+  Asmgen.compile_implementation_linear (Unit_info.prefix i.target)
+    ~progname:(Unit_info.source_file i.target)
```
here as well it looks like a single `i.target` argument would make a lot of sense for `compile_implementation_linear`.

gasche commented on file_formats/cmt_format.ml:
```diff
@@ -164,19 +164,20 @@ let record_value_dependency vd1 vd2 =
     value_deps := (vd1, vd2) :: !value_deps
 
-let save_cmt filename modname binary_annots sourcefile initial_env cmi shape =
+let save_cmt dest binary_annots initial_env cmi shape =
```
`target` or `dest`?

gasche commented on typing/persistent_env.ml:
```diff
@@ -254,7 +254,9 @@ let check_pers_struct penv f ~loc name =
+let read penv f a =
+  let modname = Unit_info.artifact_modname a in
+  let filename = Unit_info.filename a in
   snd (read_pers_struct penv f true modname filename)
```
I would expect the argument change to be pushed into `read_pers_struct`.

gasche reviewed (commented):
I did a first of review, up to the file `typing/typemod.ml` excluded.

I like the new API much better, thanks! Also, not to boast, but while reading the rest of the PR -- the many changes in the compiler codebase -- I realized that while the module takes care of a niche concern, it is actually used all over the place, so it is in fact rather important that the interface and function name be approachable.

gasche commented on parsing/unit_info.mli:
```diff
@@ -38,14 +38,14 @@ val prefix: t -> string
+   not valid according to {!is_unit_name}[ ~strict:true].
 *)
-val make: source_file:string -> string -> t
+val make: ?check_modname:bool -> source_file:string -> string -> t
```
I'm looking at the code commit-per-commit, and the label name is `check_modname` in the interface and `check` in the documentation comment. (I prefer `check_modname`.)

gasche replied:
... I saw that this was fixed later. (We should have a linter that catches this issue.)

gasche reviewed (approved):
I am happy with the final result. What do you want to do about the history?

(I think that it would make sense to squash, but you could also split into two commits, one that introduces unit_info and one with all the client-side changes at once.)

Octachron committed:
Unit_info module: metadata for compilation units and artifacts

Octachron committed:
use Unit_info everywhere

Octachron committed:
update Changes

Octachron committed:
Only add valid modname in the persistent env

Octachron commented:
I kept two commits one for introducing the new module,  the second one for using it. While rebasing https://github.com/ocaml/ocaml/pull/11736 on top of this PR, I couldn't resist the temptation to cherry-pick one simplification: with the last commit, we no longer add cmi with invalid modname in the persistent environment.
